### PR TITLE
feat: add folding builder

### PR DESCRIPTION
**Pull request by CxdInitial**
Add folding builder for:
* Enum
* Senum
* Struct
* Union
* Exception
* Service
* XsdAttrs
* ConstList
* ConstMap
* BlockComment

### CODE REVIEW

**Comment by CxdInitial** [comment]
@arhont375 
Please check this PR when you are free. 👍
No contribution process found in repo. If there is anything else that needs to be done, please let me know.

File: thrift/src/test/java/com/intellij/plugins/thrift/folding/ThriftFoldingBuilderTest.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  }
+
+  protected void doTest() throws Throwable {
+    myFixture.configureByFiles();
+    String file = getTestName(true) + ".thrift";
+    myFixture.configureByFile(file);
+    myFixture.testFolding(getTestDataPath() + "/" + file);
+  }
+
+  @Test
```
Comment (arhont375) [inline thrift/src/test/java/com/intellij/plugins/thrift/folding/ThriftFoldingBuilderTest.java]: This one also can be omitted

File: thrift/src/main/java/com/intellij/plugins/thrift/lang/ThriftFoldingBuilder.java [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    public ThriftFoldingRecursiveElementWalkingVisitor(List<FoldingDescriptor> descriptors) {
+      this.descriptors = descriptors;
+    }
+
+    @Override
+    public void visitElement(@NotNull PsiElement element) {
+      super.visitElement(element);
+
+      if (element instanceof ThriftEnum) {
+        this.visitThriftEnum((ThriftEnum) element);
```
Comment (arhont375) [inline thrift/src/main/java/com/intellij/plugins/thrift/lang/ThriftFoldingBuilder.java]: Why not to call `visitElementWithCurlyBraceChildren` directly for all elements that have curly braces?

**Review by arhont375 — COMMENTED** [review]
Thanks a lot for this improvements. Just 2 small comments

**Comment by CxdInitial** [comment]
According to the comments, I has changed the code😁.

By the way, what do you think about adding a formatter like https://github.com/thrift-labs/thrift-fmt?

File: thrift/src/test/resources/folding/foldingTwice.thrift [hunk 3]
```diff
@@ -0,0 +1,7 @@
+struct SomeStruct1 <fold text='{...}'>{
+    1: i32 someField1
+    3: i32 someField2
+    5: i32 someField3 xsd_optional xsd_nillable xsd_attrs<fold text='{...}'>{1: string someAttr}</fold>
+}</fold>
+
+const list<SomeStruct1> requestList = <fold text='[...]'>[<fold text='{...}'>{"someField1": "value1"}</fold>, <fold text='{...}'>{"someField1": "value2"}</fold>]</fold>;
```
Comment (arhont375) [inline thrift/src/test/resources/folding/foldingTwice.thrift]:
```suggestion
const list<SomeStruct1> requestList = <fold text='[...]'>[<fold text='{...}'>{"someField1": "value1"}</fold>, <fold text='{...}'>{"someField1": "value2"}</fold>]</fold>;

```

**Comment by arhont375** [comment]
> By the way, what do you think about adding a formatter like https://github.com/thrift-labs/thrift-fmt?

I don't mind having PR for this :slightly_smiling_face: 
Seems there is support for defining external formatters by plugins: https://plugins.jetbrains.com/docs/intellij/code-formatting.html#external-code-formatter

**Review by arhont375 — APPROVED** [review]

**Pull request merged**